### PR TITLE
refactor: rewrite PoKeysHal TDD and audit prompts

### LINKED ISSUE #143: Rewrite TDD compile and repository-audit prompts for PoKeysHal
## Scope

Implement Phase 2 of the audit plan in #140.

Phase 1 was completed by #141 and PR #142.

## Goal

Rewrite the two active repository prompts so they describe PoKeysHal's actual C99, PoKeysLib, LinuxCNC HAL, real-time, asynchronous, and HIL-aware workflows instead of generic web-application development.

## Files in scope

- `.github/prompts/tdd-compile.prompt.md`
- `.github/prompts/repository-audit.prompt.md`

### `tdd-compile.prompt.md`

Rewrite the prompt as a focused entry point for approved PoKeysHal implementation work.

It must:

- route implementation work to the `TDDDriver` agent;
- preserve explicit Red-Green-Refactor;
- require an existing baseline before editing;
- require Red to fail for the intended reason;
- use the smallest complete change rather than a one-check workaround;
- verify outward through the applicable repository layers;
- preserve issue and acceptance-criterion traceability;
- route hardware-dependent behavior to the existing `hil-tdd` skill;
- use `.github/skills/hil-tdd/references/result-schema.md` for HIL terminology;
- report exact commands, results, unavailable checks, and remaining risks.

Repository examples may cover:

- C parser or conversion logic;
- PoKeys protocol command and response handling;
- asynchronous mailbox, timeout, retry, or stale-response behavior;
- LinuxCNC HAL pin export and userspace smoke tests;
- RT-reachable execution paths;
- registered HIL fixtures.

### `repository-audit.prompt.md`

Rewrite the prompt as a focused audit workflow for the actual PoKeysHal repository.

It must assess:

- issue and acceptance-criterion traceability;
- C99 and PoKeysLib structure;
- asynchronous infrastructure/subsystem/integration-shell boundaries;
- LinuxCNC HAL exports and component integration;
- userspace, RT, protocol, asynchronous, and HIL verification layers;
- RT-path boundedness and unsupported RT-safety claims;
- protocol-specification evidence;
- HIL fixture registration and verified/runnable gating;
- documentation consistency and duplicate active guidance;
- exact evidence and unavailable verification.

It must not assume:

- web applications;
- authentication or JWT workflows;
- REST APIs;
- browser or UI test flows;
- databases;
- `package.json`;
- Java, TypeScript, Jest, Cypress, or Playwright;
- generic percentage-based coverage targets;
- cloud deployment or infrastructure-as-code requirements.

## Customization boundaries

- Prompts are narrow workflow entry points.
- Agents own role behavior.
- Instructions own mandatory invariants.
- Skills own detailed repeatable procedures.
- Do not duplicate the complete TDDDriver, TestingSpecialist, or `hil-tdd` content inside the prompts.

## Line budgets

- `tdd-compile.prompt.md`: preferred 80–140 lines, maximum 180.
- `repository-audit.prompt.md`: preferred 100–160 lines, maximum 220.

Do not add filler merely to reach the preferred range.

## Acceptance criteria

- Both prompts are PoKeysHal-specific.
- No generic web/auth examples remain.
- The prompts reference the actual repository verification layers.
- TDDDriver and TestingSpecialist responsibilities remain distinct.
- Hardware-dependent work routes to the approved HIL baseline.
- No redesign of #138/#139 is introduced.
- Both files remain within their maximum line budgets.
- Only the two approved prompt files are changed.

## Verification

```bash
wc -l \
  .github/prompts/tdd-compile.prompt.md \
  .github/prompts/repository-audit.prompt.md

if grep -nE \
  'TypeScript|JWT|REST|database|browser|Java|Jest|Cypress|Playwright|bcrypt|jsonwebtoken|package\.json|npm test' \
  .github/prompts/tdd-compile.prompt.md \
  .github/prompts/repository-audit.prompt.md
then
    echo "ERROR: generic stack content remains" >&2
    exit 1
else
    echo "PASS: no generic stack content found"
fi

git diff --name-only origin/main...HEAD
git diff --check
git diff --ignore-space-at-eol
```

Expected changed files:

```text
.github/prompts/tdd-compile.prompt.md
.github/prompts/repository-audit.prompt.md
```

## Out of scope

* Agent changes.
* Instruction or skill changes.
* HIL fixture changes.
* Test implementation.
* Hardware execution.
* Production C or HAL changes.
* Phase 3 consolidation.

## Traceability

Part of #140.

### PR DESCRIPTION
**Pull request by zarfld**
Closes #143
Part of #140

## Changes

Rewrites both active repository prompts to describe PoKeysHal-specific C99, PoKeysLib, LinuxCNC HAL, real-time, asynchronous, and HIL-aware workflows. Removes all generic web-application content (TypeScript, JWT, REST, database, browser, Jest, Cypress, etc.).

- `tdd-compile.prompt.md`: 801 → 65 lines. Narrow orchestration contract routing to TDDDriver; requires baseline, Red-Green-Refactor through applicable layers, hil-tdd routing for hardware-dependent criteria, and full handoff evidence including HIL-observed and partially-verified status.
- `repository-audit.prompt.md`: 1899 → 150 lines. Read-only audit workflow covering traceability (StR/REQ-F/REQ-NF/ADR/ARC-C/QA-SC/TEST with acceptance-criterion mapping), C99/PoKeysLib structure, async boundaries, HAL exports, RT constraints (including `free` on RT paths), protocol evidence, HIL fixture gating (runnable+verified+setup ID+revision), documentation consistency, and exact-evidence output format with partially-verified disclosure.

## Verification evidence

```
$ wc -l .github/prompts/tdd-compile.prompt.md .github/prompts/repository-audit.prompt.md
  65 .github/prompts/tdd-compile.prompt.md
 150 .github/prompts/repository-audit.prompt.md
 215 total
```

```
$ git diff --name-only origin/main...HEAD
.github/prompts/repository-audit.prompt.md
.github/prompts/tdd-compile.prompt.md
```

```
$ git diff --check origin/main...HEAD
(no output — PASS)
```

```
$ grep -nE 'TypeScript|JWT|REST|database|browser|Java|Jest|Cypress|Playwright|bcrypt|jsonwebtoken|package\.json|npm test' \
    .github/prompts/tdd-compile.prompt.md .github/prompts/repository-audit.prompt.md
(no output — PASS)
```

**Frontmatter**: both files use `agent` + `argument-hint`; `mode` and `applyTo` absent.

**Prompt vs agent/skill boundary**: `tdd-compile` is an orchestration contract only — it does not reproduce TDDDriver's procedure steps. `repository-audit` references `result-schema.md` for HIL definitions; it does not redefine them.

**VS Code prompt invocation**: not performed. The prompts were not run from VS Code after amendment; invocation behavior in the VS Code prompt picker is unverified.

### CODE REVIEW

**Review by zarfld — COMMENTED** [review]
## Independent Phase 2 review: changes required before merge

I reviewed the complete final contents of both prompt files against issue #143, `AGENTS.md`, the engineering-discipline and C/RT instructions, the rewritten TDDDriver/TestingSpecialist agents, and the approved `hil-tdd` skill/result schema.

### What passes

- Branch topology is correct: one commit ahead of `main`, zero behind.
- Exactly the two files authorized by #143 are changed.
- Both files are within their line budgets.
- Generic web/auth stack material is removed.
- The prompts use PoKeysHal-specific C99, PoKeysLib, HAL, async, RT, protocol, and HIL terminology.
- HIL fixture gating and the prohibition on userspace/mock `HIL-verified` claims are directionally correct.

### Required changes

1. **Correct both prompt-file frontmatter blocks.** Current prompt-file metadata uses `agent`, not `mode`; `applyTo` is instruction-file metadata and does not scope or recommend a `.prompt.md` file. Use `agent: TDDDriver` for the implementation prompt and an appropriate agent value for the audit prompt, then remove `applyTo`.
2. **Reduce TDDDriver duplication.** `tdd-compile.prompt.md` currently reproduces nearly the complete TDDDriver workflow. Keep it as a concise invocation/orchestration contract: input issue, baseline gate, explicit Red–Green–Refactor sequence, outward verification, HIL routing, and required handoff. The agent remains the authority for implementation behavior.
3. **Remove unsupported audit requirements.** The canonical C/RT contract prohibits dynamic allocation, blocking operations, and unbounded execution on RT-reachable paths. It does not categorically prohibit all VLAs/non-constant bounds, all cross-subsystem calls, or require every exported pin to update on every HAL invocation. Audit against the exact canonical responsibility and boundedness rules rather than inventing stricter rules.
4. **Clarify read-only evidence collection.** The audit must inspect existing command output/CI/log evidence, not execute the listed build and HAL commands. Missing evidence should be reported as unavailable.
5. **Do not duplicate the result-schema table.** `result-schema.md` explicitly declares itself the single authority and says not to reproduce its table elsewhere. Reference it and require its terms; remove the copied HIL definitions. In the TDD handoff, include `HIL-observed` where applicable and explicitly require `partially verified` when a required verification layer is unavailable.
6. **Refresh PR verification evidence after correction.** Include exact changed files, final line counts, current prompt-frontmatter validation, forbidden-term scan, `git diff --check`, and `git diff --ignore-space-at-eol`.

The existing Copilot comment about adding `.comp` to `applyTo` should not be implemented: prompt files are manually invoked; `applyTo` belongs to `.instructions.md` files. Removing `applyTo` is the correct fix.

File: .github/prompts/tdd-compile.prompt.md [hunk 1]
```diff
@@ -1,801 +1,116 @@
 ---
 mode: agent
```
Comment (zarfld) [inline .github/prompts/tdd-compile.prompt.md]: This is instruction-file metadata, not current prompt-file metadata. Replace `mode: agent` with `agent: TDDDriver` so the prompt actually routes to the custom implementation agent, and remove the `applyTo` block. Prompt files are manually invoked; `applyTo` is defined for `.instructions.md`, not `.prompt.md`.

File: .github/prompts/tdd-compile.prompt.md [hunk 2]
```diff
@@ -1,801 +1,116 @@
...
-# TDD Compile Prompt (GitHub Issues)
+# TDD Compile – PoKeysHal Implementation Workflow
 
-You are a Test-Driven Development (TDD) specialist enforcing **ISO/IEC/IEEE 12207:2017** and **Extreme Programming (XP) best practices**.
+Entry point for implementing an approved GitHub issue through Red-Green-Refactor.
+Implementation behavior is owned by the **TDDDriver** agent; invoke it with the issue number.
+This prompt defines the required sequence, evidence contract, and routing.
 
-## 🎯 Core Workflow: GitHub Issues + TDD Cycle
+## Prepare
```
Comment (zarfld) [inline .github/prompts/tdd-compile.prompt.md]: From here through the handoff section, this prompt reproduces most of `tdd-driver.agent.md` rather than remaining a narrow entry point. Compress this to orchestration and acceptance gates, then rely on `agent: TDDDriver` and the canonical instructions for the detailed implementation procedure.

File: .github/prompts/tdd-compile.prompt.md [hunk 3]
```diff
@@ -1,801 +1,116 @@
...
----
+## Handoff evidence
 
-**Remember**: Tests ALWAYS come first! Red → Green → Refactor → PR with issue links! 🔴🟢🔨
+Report exactly:
+- files changed and requirement addressed;
+- tests added or changed and commands executed with results;
+- unavailable checks and why;
+- remaining risks;
+- precise status from:
```
Comment (zarfld) [inline .github/prompts/tdd-compile.prompt.md]: This closed status list omits `HIL-observed` even though the prompt cites the shared schema, and it omits the engineering contract's required `partially verified` state when a required layer is unavailable. Add those semantics without redefining the schema terms.

File: .github/prompts/repository-audit.prompt.md [hunk 4]
```diff
@@ -1,1900 +1,136 @@
...
-- Track compliance score trends
-- Monitor migration progress
-- Alert on regression in key metrics
-
----
-
-**Know where you stand, plan where you're going!** 📊
\ No newline at end of file
+---
+mode: agent
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md]: Use prompt-file frontmatter here as well: replace `mode: agent` with `agent: agent` (or a deliberately selected custom agent) and remove `applyTo`. The current fields do not provide path-based prompt activation.

File: .github/prompts/repository-audit.prompt.md [hunk 5]
```diff
@@ -1,1900 +1,136 @@
...
+- Confirm file responsibilities match `c-architecture-realtime.instructions.md`.
+- Flag VLA, `malloc`, `calloc`, `realloc`, or unbounded loops on any RT-reachable path.
+- Flag use of prohibited C features (dynamic allocation, blocking calls, non-constant
+  bounds) in files included on RT-reachable call paths.
+
+### 3. Async infrastructure / subsystem / integration-shell boundaries
+
+- `PoKeysLibAsync.c`: mailbox and dispatch infrastructure only — no subsystem logic.
+- `PoKeysLib*Async.c`: per-subsystem HAL export, send, parse, and optional scheduler
+  registration only — no cross-subsystem calls, no transport logic.
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md]: This block introduces restrictions not stated by the canonical C/RT contract: all VLAs/non-constant bounds and all cross-subsystem calls are not categorically prohibited. Audit the actual rules—no dynamic allocation, blocking calls, or unbounded execution on RT-reachable paths, plus the exact file responsibilities from `c-architecture-realtime.instructions.md`—without adding stricter requirements.

File: .github/prompts/repository-audit.prompt.md [hunk 6]
```diff
@@ -1,1900 +1,136 @@
...
+- `experimental/pokeys_async.c`: integration shell only — no business logic.
+- Report any cross-boundary leakage as a finding with file and line reference.
+
+### 4. LinuxCNC HAL exports and component integration
+
+- HAL-visible struct members must use `hal_u32_t`, `hal_s32_t`, `hal_bit_t`, or
+  `hal_float_t`. Flag any raw C type used where a HAL type is required.
+- Confirm pin names follow the documented naming convention.
+- Confirm exported pins are updated on every HAL function invocation.
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md]: “Updated on every HAL function invocation” is stronger than the repository contract and may be false for asynchronously refreshed data. The canonical requirement is that exported pins have a real, reachable update path and are updated according to the subsystem/component contract—not necessarily on every invocation.

File: .github/prompts/repository-audit.prompt.md [hunk 7]
```diff
@@ -1,1900 +1,136 @@
...
+- Confirm the userspace and RT component entry points match the integration shell.
+
+### 5. Verification layers
+
+Assess what evidence exists at each layer:
+
+| Layer | How assessed |
+|---|---|
+| Compile and link | `bash test_compile.sh`, `make -f Makefile.noqmake` results |
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md]: This conflicts with the read-only constraint unless it explicitly means existing recorded output. Reword the column/row to require inspection of existing CI logs, PR evidence, or saved command results; when none exists, classify the layer as unavailable rather than running the commands.

File: .github/prompts/repository-audit.prompt.md [hunk 8]
```diff
@@ -1,1900 +1,136 @@
...
+| Term | Meaning |
+|---|---|
+| Implemented | Production behavior exists in source |
+| Compiled | Compilation succeeded only |
+| Tested | Named tests executed and passed |
+| HIL-observed | Exploratory observation recorded as oracle evidence |
+| HIL-test-executed | Named HIL test ran to completion |
+| HIL-verified | Named HIL test passed on a named fixture at a stated revision |
+| RT-validated | Tested in the applicable LinuxCNC RT environment |
+| Timing-validated | Measured against a documented threshold with an identified method |
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md:122]: `result-schema.md` says it is the single authority and that its table must not be duplicated elsewhere. Remove the copied HIL definitions and link to the schema for those terms. Keep only repository-level non-HIL claim guidance that is not already defined by the schema.

**Review by zarfld — COMMENTED** [review]
## Second independent review: improved, but not yet merge-ready

I verified the amended head rather than relying on the Copilot session summary. The following claims are correct:

- the branch is one commit ahead of `main` and changes only the two approved prompt files;
- `agent` and `argument-hint` are valid prompt-file fields;
- `mode` and `applyTo` have been removed;
- the generic web/auth stack is gone;
- the TDD prompt is now materially smaller and distinct from `TDDDriver`;
- the read-only evidence table no longer instructs the auditor to execute builds;
- the copied HIL term definitions were removed;
- `HIL-observed` and the `partially verified` state were added.

The statement that **all findings are addressed** is still too strong. Please correct these remaining items:

1. **Do not invent a VLA prohibition.** `repository-audit.prompt.md` still says to flag every variable-length array on an RT-reachable path. The canonical C/RT contract prohibits dynamic allocation, blocking operations, and unbounded execution, but does not categorically prohibit every bounded VLA. Remove `variable-length arrays` from that rule unless the canonical instruction is separately changed to establish that invariant.

2. **Make outward verification explicitly conditional.** In `tdd-compile.prompt.md`, say “through the applicable layers” and mark userspace HAL, RT, and HIL layers as applicable/required by the acceptance criteria. The current wording can be read as requiring RT and HIL for every implementation issue, which contradicts the agent and issue #143 wording.

3. **Complete the traceability audit vocabulary.** The canonical root instructions include `QA-SC`, but the audit’s issue-type inventory omits it. Add `QA-SC` so the audit does not report against an incomplete repository issue model.

4. **Complete HIL fixture registration checks.** The audit correctly requires `runnable: true` and `fixture_status: verified`, but issue #143 also requires fixture registration. Require recorded evidence to identify the registered setup ID and revision; do not duplicate the detailed preflight procedure. Also call `HIL-observed`, `HIL-test-executed`, and `HIL-verified` **status terms**, not “outcome vocabulary”; PASS/FAIL/ERROR/SKIPPED are the outcomes.

5. **Require exact evidence in the output.** The output format currently asks only for a file or issue reference and observed/required state. Add an exact evidence field such as file:line, issue/PR reference, CI run/log, test report, or HIL artifact. Also state that an audit with a required unavailable layer is `partially verified`.

6. **Do not restrict corrective issue types incorrectly.** The corrective-plan output currently limits recommendations to `REQ-F`, `TEST`, `ARC-C`, or process change. Findings may require `REQ-NF`, `ADR`, `QA-SC`, bug, integration, documentation, or another existing repository issue type. Use “appropriate existing issue type” with examples rather than a closed four-type list.

7. **Refresh the PR description and verification evidence.** The PR body still reports the pre-amendment 116/136 line counts and describes the old file shape. Update it to the current 65/135 counts and include the frontmatter validation and `git diff --check` result. The only Actions runs on this head are unrelated lifecycle-issue workflows; they do not validate these prompt files. State that actual VS Code prompt discovery/invocation was not executed if it was not tested.

After those corrections, resolve the obsolete review threads and request the final pass. The current content is close, but these items are required for issue #143’s no-invented-rules, exact-evidence, traceability, and HIL-registration acceptance criteria.

File: .github/prompts/repository-audit.prompt.md [hunk 9]
```diff
@@ -1,1900 +1,148 @@
...
+
+## Audit areas
+
+### 1. Issue and acceptance-criterion traceability
+
+- Confirm issue types StR, REQ-F, REQ-NF, ADR, ARC-C, QA-SC, TEST exist with labels.
+- Confirm REQ issues link upward to StR; TEST issues carry `Verifies: #N`.
+- Confirm source uses `/* Implements: #N (REQ-F-xxx) */`; tests use `/* Verifies: #N */`.
+- Confirm PRs use `Fixes #N` or `Implements #N`.
+- Flag orphaned requirements, untraced tests, and PRs without an issue link.
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md:30]: This section still does not test the acceptance-criterion link promised by its heading and required by #143. Add a compact rule such as: “For each in-scope approved acceptance criterion, identify the implementing artifact and verification evidence; flag missing or ambiguous links.” Source/test comment conventions alone do not prove criterion coverage.

File: .github/prompts/repository-audit.prompt.md [hunk 10]
```diff
@@ -1,1900 +1,148 @@
...
+
+### 2. C99 and PoKeysLib structure
+
+- Confirm file responsibilities match `c-architecture-realtime.instructions.md`.
+- Flag dynamic allocation (`malloc`, `calloc`, `realloc`), blocking calls, or
```
Comment (zarfld) [inline .github/prompts/repository-audit.prompt.md]: Add `free` to the explicit RT-path scan. `.github/instructions/c-architecture-realtime.instructions.md` prohibits `malloc`, `free`, and dynamic allocation on RT-reachable paths; omitting `free` can miss a real RT-path violation.

**Review by zarfld — COMMENTED** [review]
## Third independent review

The amended files now resolve most findings from review `4829158807`: prompt frontmatter is valid, outward verification is conditional, QA-SC is present, HIL registration/status/outcome wording is corrected, exact evidence is required, and the corrective-plan issue types are non-exhaustive.

Three corrections remain before merge:

1. Add explicit acceptance-criterion coverage to the audit. The section is titled “Issue and acceptance-criterion traceability,” but it currently verifies issue/link/comment conventions only. Require the audit to confirm that each in-scope approved acceptance criterion has corresponding implementation and verification evidence, or record the missing link as a finding.

2. Include `free` in the RT-path memory-operation scan. The canonical C/RT instruction prohibits `malloc`, `free`, and dynamic allocation on RT-reachable paths; the current explicit function list omits `free`.

3. The PR body update did not take effect on GitHub. The remote PR still reports 116/136 lines and the old verification block. Update it to the actual 65/148 evidence and disclose that VS Code prompt discovery/invocation was not performed. Verify the result with `gh pr view 145 --json body` after editing.

After those edits, resolve the addressed/outdated review threads and rerun the final two-file, line-count, forbidden-term, frontmatter, and `git diff --check` evidence. No runtime, RT, or HIL execution is required for this documentation-only change.

**Review by zarfld — COMMENTED** [review]
## Final review on current head `d12bf6c`

The two findings from review `4829303236` are resolved on this head:

- acceptance-criterion mapping now requires an implementing artifact and verification evidence for each in-scope approved criterion;
- the RT-path scan now explicitly includes `free`.

The remote PR body is also current: it reports 65 and 150 lines, the two-file scope, `git diff --check`, the forbidden-term scan, frontmatter validation, responsibility-boundary validation, and the disclosed absence of a VS Code invocation test.

The implementation now satisfies issue #143's scope and acceptance criteria. No further content changes are required before merge. Resolve the addressed review threads so the conversation state matches the current files.

Issue #140 remains open for Phase 3 consolidation after this PR merges.
